src/parser/parser.cc

```diff
@@ -661,15 +754,21 @@ class Parser {
           Consume(TokenType::kDefn);
           auto global_tok = Match(TokenType::kGlobal);
           auto global_name = global_tok.ToString();
-          auto global = GlobalVar(global_name);
-          try {
-            global_names.Add(global_name, global);
-          } catch (const DuplicateKeyError& e) {
-            this->diag_ctx->Emit(Diagnostic::Error(global_tok->span) << "a function with the name "
-                                                                     << "`@" << global_name << "` "
-                                                                     << "was previously defined");
-          }
-          auto func = ParseFunctionDef();
+          auto global = AddOrGet(&global_names, global_name);
+          auto func = WithSpan<relay::Function>([&]() {
+            // try {
```

Comment:
Fix the error handling around this, was blanket catching exceptions earlier which is an issue.